elytron/WFLY-14984-encrypted-expression-for-client-config.adoc

```diff
@@ -0,0 +1,172 @@
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-14984[WFLY-14984]
+
+// === Related Issues
+
+// * https://issues.redhat.com/browse/WFCORE[WFCORE-XXXX]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [x] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+* The ability to specify attribute values as encrypted expressions should be added to wildfly-config.xml. 
+
+* The parsing needs to be able to run separately from the application server, as the server may not be running while the client configuration is being set up. 
+
+* Configurations may not depend on components created for server configuration, such as `resolvers` or `elytron/expression`. The resolver for the encrypted expression should be configured through the client configuration. 
+
+* The `encrypted-expression` would be a new configuration block added to the client config under the new schema. The new attributes added include `resolvers`, with attributes called `resolver` under it. Additionally, <credential-stores> attribute would be added to the schema with the <credential-store> attributes added to it. The structures of this attribute would be the same as the one one under the `authentication-client` schema. 
+
+** There may be multiple `resolver` attribute listed under <resolvers>, and there may be multiple <credential-store> attributes added under <credential-stores>.
+
+** The configuration for this block can be specified as follows: 
+```
+<encrypted-expression>
+	<credential-store name="credential-store-name" type="PKCS12" provider="provider-name"/>
+        <attributes>
+            <attribute name="path" value="path/to/credential-store-name.cs" />
+        </attributes>
+        <protection-parameter-credentials>
+            <clear-password password="password" />
+        </protection-parameter-credentials>
+    </credential-store>
+    <resolvers>
+        <resolver name= “resolver-name” credential-store= “credential-store-name key=“key”/>
+        <resolver name= “resolver-name-2” credential-store= “credential-store-name key=“example”/>
+    </resolvers>
+    <default-resolver name=resolver-name>
```

Comment:
Does default-resolver need to be its own element? Can it be an attribute defined on the expression-resolver like in the server side config? e.g., see the xml config here:

https://wildfly-security.github.io/wildfly-elytron/blog/wildfly-encrypted-expressions/

Reply:
if it is not added as a separate element, would we add it as a property for the encrypted expression element? 
so like 
```
<encrypted-expression default-resolver="resolver-name">
INSERT CONFIGURATION
</encrypted expression>
```
If we do go with this, I am not sure if we can ensure that `resolver-name` is one of the resolvers specified under the <expression-resolver> elements. I think the parser currently checks to make sure it is.

Reply:
Just looking at the example from the blog post:

```
<expression-resolver default-resolver="main-resolver">
    <resolver name="initial-resolver" credential-store="initial" secret-key="key"/>
    <resolver name="main-resolver" credential-store="main" secret-key="main-key"/>
</expression-resolver>
```

Can we make use of this format?

Reply:
Yes, I will update the schema and parser to use this. 
Also, would it be better to make the default-resolver a required attribute? or make it optional, but if the user does not specify one, then the first resolver becomes the default resolver?

Reply:
How is this done on the server side?

Reply:
The server does neither. 
When adding an expression resource using the following command: 
`/subsystem=elytron/expression=encryption:add(resolvers=[{name=initial-resolver, credential-store=initial, secret-key=key}])` defining a value for the `default-resolver` attribute is optional on the server side. 
Now if someone tries to create an expression without defining a resolver, then the output is as follows: 
```
/subsystem=elytron/expression=encryption:create-expression(clear-text=password)
{
    "outcome" => "failed",
    "failure-description" => "WFLYELY01200: The name of the resolver to use was 
not specified and no default-resolver has been defined.",
    "rolled-back" => true
}
```

On the client, the create-expression part is done using the `--encrypt` option on the wildfly-elytron tool, which would give you an output like this: 
```
Clear text encrypted to token 'RUxZAUMQQ1PVWZuimKZCpt0s6MONrl7cbqLGPhyBBmm83fXdfZA=' using alias 'key'.
```

It would be up to the user to convert it to something like `{ENC::resolver_name:RUxZAUMQQ1PVWZuimKZCpt0s6MONrl7cbqLGPhyBBmm83fXdfZA=}` (where the expression resolver called `resolver-name` would be used to decrypt)
or they can convert it to `{ENC::RUxZAUMQQ1PVWZuimKZCpt0s6MONrl7cbqLGPhyBBmm83fXdfZA=}` (where the default expression resolver would be used to decrypt)

There is code in place already in the resolver class to handle cases where the user goes with the second configuration, but there is no default-expression-resolver defined. So, if we do want to do something similar to what the server does, I think it may be possible to throw an exception when trying to resolve the expression. However, since we are using the wildfly-elytron tool to encrypt, we cannot throw the same exception when creating the expression, which is what the server does.

Reply:
Ok given that, I think it would be good to have the default-resolver attribute be optional here too.

An encrypted expression provided in wildfly-config should either specify the resolver name or rely on the default being used.

If relying on the default and the default-resolver hasn't been defined, then that's invalid configuration and an exception should occur when attempting to resolve it.